packages/web/src/components/cells/code.tsx

```diff
@@ -270,609 +334,54 @@ export default function CodeCell(props: Props) {
-    }
-    const file_response = await getFileContent(response.file);
-    if (file_response.result.type === 'cell') {
-      document
-        .getElementById(file_response.result.filename)
-        ?.scrollIntoView({ behavior: 'smooth' });
-    } else {
-      if (file_response.error) {
-        console.error('Error fetching file content:', file_response.result);
-      } else {
-        editor(file_response.result.content);
+  async function onGetDefinitionContents(pos: number, cell: CodeCellType): Promise<string> {
+    return new Promise((resolve, reject) => {
+      if (!channel) {
+        return;
       }
-    }
-  }
-
-  console.log({ location: mapCMLocationToTsServer(cell.source, pos) });
-  channel.on('tsserver:cell:definition_location:response', gotoDefCallback);
-  channel.push('tsserver:cell:definition_location:request', {
-    sessionId: session.id,
-    cellId: cell.id,
-    request: { location: mapCMLocationToTsServer(cell.source, pos) },
-  });
-}
-
-type CodeEditorBaseProps = {
-  cell: CodeCellType;
-  session: SessionType;
-};
-type CodeEditorRegularProps = CodeEditorBaseProps & {
-  readOnly?: false;
-  channel: SessionChannel;
-  runCell: () => void;
-  formatCell: () => void;
-  updateCellOnServer: (cell: CodeCellType, attrs: CodeCellUpdateAttrsType) => void;
-};
-type CodeEditorReadOnlyProps = CodeEditorBaseProps & {
-  readOnly: true;
-  channel?: SessionChannel;
-  runCell?: () => void;
-  formatCell?: () => void;
-  updateCellOnServer?: (cell: CodeCellType, attrs: CodeCellUpdateAttrsType) => void;
-};
-type CodeEditorProps = CodeEditorRegularProps | CodeEditorReadOnlyProps;
-
-function CodeEditor({
-  readOnly,
-  channel,
-  cell,
-  session,
-  runCell,
-  formatCell,
-  updateCellOnServer,
-}: CodeEditorProps) {
-  const { theme, codeTheme } = useTheme();
 
-  const {
-    updateCell: updateCellOnClient,
-    getTsServerDiagnostics,
-    getTsServerSuggestions,
-  } = useCells();
-
-  const updateCellOnServerOrNoop = useCallback<NonNullable<typeof updateCellOnServer>>(
-    (cell, attrs) => {
-      if (!updateCellOnServer) {
-        return;
+      async function gotoDefCallback({ response }: TsServerDefinitionLocationResponsePayloadType) {
+        if (!channel) {
+          return;
+        }
+
+        channel.off('tsserver:cell:definition_location:response', gotoDefCallback);
+        if (response === null) {
+          reject(new Error(`Error fetching file content: no response!`));
+          return;
+        }
+        const file_response = await getFileContent(response.file);
+        if (file_response.result.type === 'cell') {
+          document
+            .getElementById(file_response.result.filename)
+            ?.scrollIntoView({ behavior: 'smooth' });
+        } else {
+          if (file_response.error) {
+            reject(new Error(`Error fetching file content: ${file_response.result}`));
+          } else {
+            resolve(file_response.result.content);
+          }
+        }
```

Comment:
I wanted to call out a pretty substantial rework I did of the go to definition here as part of an earlier abstration attempt, but I left it in because it seemed to be a bit cleaner.

In summary, make `onGetDefinitionContents(...)` return a `Promise<string>`, rather than call an ad-hoc callback.